src/aws-cpp-sdk-core/source/client/RequestCompression.cpp

```diff
@@ -0,0 +1,229 @@
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+#include <aws/core/client/RequestCompression.h>
+#include <aws/core/utils/logging/LogMacros.h>
+
+#ifdef ENABLED_ZLIB_REQUEST_COMPRESSION
+#include "zlib.h"
+
+#if defined(MSDOS) || defined(OS2) || defined(WIN32) || defined(__CYGWIN__)
+#include <fcntl.h>
+#include <io.h>
+#define SET_BINARY_MODE(file) setmode(fileno(file), O_BINARY)
+#else
+#define SET_BINARY_MODE(file)
+#endif
+// Defining zlib chunks to be 256k
+#define ZLIB_CHUNK 263144
+#endif
+
+static const char AWS_REQUEST_COMPRESSION_LOG_TAG[] = "RequestCompression";
+#ifdef ENABLED_ZLIB_REQUEST_COMPRESSION
+static const char AWS_REQUEST_COMPRESSION_ALLOCATION_TAG[] =
+    "RequestCompressionAlloc";
+#endif
+
+Aws::String
+Aws::Client::GetCompressionAlgorithmId(const CompressionAlgorithm &algorithm) {
+  switch (algorithm) {
+  case CompressionAlgorithm::GZIP:
+    return "gzip";
+  default:
+    return "";
+  }
+}
+
+Aws::Utils::Outcome<std::shared_ptr<Aws::IOStream>, bool>
+Aws::Client::RequestCompression::compress(
+    std::shared_ptr<Aws::IOStream> input,
+    const CompressionAlgorithm &algorithm) const {
+
+#ifdef ENABLED_ZLIB_REQUEST_COMPRESSION
+  if (algorithm == CompressionAlgorithm::GZIP) {
+    // calculating stream size
+    input->seekg(0, input->end);
+    auto streamSize = input->tellg();
+    input->seekg(0, input->beg);
+
+    AWS_LOGSTREAM_TRACE(AWS_REQUEST_COMPRESSION_LOG_TAG,
+                        "Compressing request of " << streamSize << " bytes.");
+
+    // Preparing output
+    std::shared_ptr<Aws::IOStream> output = Aws::MakeShared<Aws::StringStream>(
+        AWS_REQUEST_COMPRESSION_ALLOCATION_TAG);
+
+    // Prepare ZLIB to compress
+    int ret = Z_NULL;
+    z_stream strm;
+    std::unique_ptr<unsigned char[]> in(new unsigned char[ZLIB_CHUNK]);
+    std::unique_ptr<unsigned char[]> out(new unsigned char[ZLIB_CHUNK]);
+    // Cleaning Zlib stream state flags
+    strm.zalloc = Z_NULL;
+    strm.zfree = Z_NULL;
+    strm.opaque = Z_NULL;
+
+    const int max_window_gzip = 31;
+    const int default_mem_level_usage = 8;
+    ret =
+        deflateInit2(&strm, Z_DEFAULT_COMPRESSION, Z_DEFLATED, max_window_gzip,
+                     default_mem_level_usage, Z_DEFAULT_STRATEGY);
+    if (ret != Z_OK)
+      return false;
+    // Compress
+    //  do {
+    // Fill the buffer
+    if (streamSize > ZLIB_CHUNK) {
+      // Skipping multi chunk for now
+      return false;
+    } else {
+      int flush = Z_FINISH;
+      input->read(reinterpret_cast<char *>(in.get()), streamSize);
```

Comment:
Because of disabled exceptions, we are not able to verify that read succeeds 
Please check for badbit or [fail()](https://en.cppreference.com/w/cpp/io/basic_ios/fail) after read/get is called